src/book/02-system/01-nubots/03-nuclear.mdx

```diff
@@ -5,3 +5,144 @@ title: NUClear
+
+For reactions to occur, at least one Binding DSL word should be present in the DSL Request. The binding DSL words are: `Trigger`, `With`, `Every`, `Always`, `Startup`, `Shutdown`, `TCP`, `UDP`, and `Network`.
+
+| DSL Word | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                |
+| -------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Trigger  | Is used to request data dependent reactions. `on<Trigger<T>>()` will execute the task whenever T is emitted into the system. Read only access to T is given via a callback. `on<Trigger<T1, T2, ... >>()` requires multiple types and will only trigger when all the trigger types have been emitted at least once since the last occurrence of the event.                                                                                                  |
+| With     | Is used to define any extra data which should be provided to a subscribing a reaction. This should be fused with at least one other DSL word as the use of this word alone will not trigger a reaction within the system. `on<Trigger<T1>, With<T2>>()` will trigger when T1 is emitted into the system. It will give read-only access of T1 and T2 to the callback. If T2 is not present when T1 is emitted to the system, the reaction will not run. |
+| Last     | Instructs the PowerPlant to store the last $n$ messages (of the associated type) to the cache and provide read-only access to the subscribing reaction. `on<Last<n, Trigger<T, ...>>>()` will store the next $n$ messages of each of the `Trigger` types in a list with the oldest message first. After $n$ messages are stored, the trigger of a new reaction will cause the oldest message to be dropped and the newest copy appended to the list.                                         |
+| Optional | Is used to signify any optional properties in the DSL request. `on<Trigger<T1>, Optional<With<T2>()` will run when T1 is emitted into the system. If T2 is available, then the reaction will be given read-only access to the most recent emission of T2.                                                                                                                                                                                             |
```

Comment:
Here is the explanation of how Optional works/should be used.

When you have an argument in NUClear it's not actually a const reference of that type, it is actually an std::shared_ptr<const T> object. NUClear just has the ability to work out if it could dereference an object to satisfy the type of the argument. So an on statement can look like either of these options.

```cpp
on<Trigger<DataType>>().then([this] (std::shared_ptr<const DataType> msg) {});
// OR
on<Trigger<DataType>>().then([this] (const DataType& msg) {});
```
The second one is easier to use most of the time, as when you're not using optional, NUClear won't run when you have no data.
However the times you do want to use the first one is when you're using Optional, because at those times that second shared_ptr argument could be a nullptr, and if you don't have a pointer you can't check if it's null or not. However if you have another way to check if that secondary object is null you can use the second option.

An example of the usage in the NUbots codebase is here
https://github.com/NUbots/NUbots/blob/master/module/platform/darwin/SensorFilter/src/SensorFilter.cpp#L248
and checking if it's null here
https://github.com/NUbots/NUbots/blob/master/module/platform/darwin/SensorFilter/src/SensorFilter.cpp#L314

The other time that the first option with the shared_ptr is used is when we want to save the pointer somewhere for future use as NUClear will only maintain it so long as it's the most recent object of that type. Although in that case since we use neutrons we often just use shared_from_this() instead, e.g.
https://github.com/NUbots/NUbots/blob/master/module/vision/GreenHorizonDetector/src/GreenHorizonDetector.cpp#L161
But in cases like that, we could have also just taken in a shared_ptr rather than a const reference

Reply:
I've updated it a lil bit. I'll add this example to the examples section. Let me know if there are any other particularly good examples of anything NUClear-y in the codebase that you can think of.